addon/-record-data-private/system/model/internal-model.js

```diff
@@ -1240,3 +1223,29 @@ export default class InternalModel {
     return reference;
   }
 }
+
+function assertRecordsPassedToHasMany(records) {
+  assert(`You must pass an array of records to set a hasMany relationship`, isArrayLike(records));
+  assert(
+    `All elements of a hasMany relationship must be instances of DS.Model, you passed ${inspect(
+      records
+    )}`,
+    (function() {
+      return A(records).every(record => record.hasOwnProperty('_internalModel') === true);
+    })()
+  );
+}
+
+function extractRecordDatasFromRecords(records) {
+  return records.map(extractRecordDataFromRecord);
+}
+
+function extractRecordDataFromRecord(record) {
```

Comment:
I'd recommend renaming `record` given that it's a union type.

something like `recordOrPromiseProxy`

```js
function extractRecordDataFromValue(recordOrPromiseProxy) {
  if (/* detect promise proxy case */) {
    // promise proxy
    return recordOrPromiseProxy;
  }

  // record case
  return record._internalModel._modelData;
}
```